frontend/app/services/redis-service.server.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * The RedisService is a service module for interacting with a Redis server. It
+ * provides methods for getting, setting, and deleting values from Redis.
+ *
+ * The getRedisService() function creates a singleton instance of the Redis
+ * client. The instance is lazy initialized as needed, so it will only be
+ * created when a client requests it. This ensures that a single connection pool
+ * is used for the entire application.
+ *
+ * Example usage:
+ *
+ *   const redisService = getRedisService({ url: 'redis://redis.example.com' });
+ *   await redisService.set('key', 'value', { EX: 600 }); // expire in 10 minutes
+ *   const value = await redisService.get('key');
+ *   await redisService.del('key');
+ *
+ * @see https://redis.io/docs/connect/clients/nodejs/
+ */
+import { type SetOptions, createClient } from 'redis';
+
+import { getLogger } from '~/utils/logging.server';
+
+const log = getLogger('redis-service.server');
+
+/**
+ * Singleton redis client instance; lazy initialized as needed.
+ */
+let redisClient: ReturnType<typeof createClient>;
+
+export interface RedisServiceConfigurationOptions {
+  url: string;
+  username?: string;
+  password?: string;
+}
+
+export function getRedisService(config: RedisServiceConfigurationOptions) {
+  return {
+    /**
+     * @see https://redis.io/commands/get/
+     */
+    get: async (key: string) => {
+      redisClient ??= await createRedisClient(config.url, config.username, config.password);
```

Comment:
Create this inside `getRedisService` and reuse it.

```ts
async function getRedisClient() {
 return redisClient ??= await createRedisClient(config.url, config.username, config.password);
}
```

Reply:
Fixed in latest push